main/src/ca/uwaterloo/flix/util/collection/Chain.scala

```diff
@@ -43,10 +43,81 @@ sealed trait Chain[+A] extends Iterable[A] {
     }
   }
 
+  def isEmpty: Boolean = this match {
+    case Chain.Empty => true
+    case Chain.Link(l, r) => l.isEmpty && r.isEmpty
+    case Chain.Many(cs) => cs.forall(_.isEmpty)
+    case Chain.Proxy(xs) => xs.isEmpty
+  }
+
+  def head: Option[A] = this match {
+    case Chain.Empty => None
+    case Chain.Link(l, _) => l.head
+    case Chain.Many(cs) => cs.find(_.head.isDefined).flatMap(_.head)
+    case Chain.Proxy(xs) => xs.headOption
+  }
+
+  /**
+    * Returns `this` as a [[List]].
+    */
+  final def toList: List[A] = this match {
+    case Chain.Empty => List.empty
```

Comment:
My untested theory is that this match checks whether `this == Chain.Empty`, which causes an infinite loop due to the `toList` call in `equals`.

Can't we sidestep all of this by inheriting Iterable and then just doing `Iterable.iterate.toList`?

Reply:
I would prefer not to depend on Scalas broken collections.

Reply:
> My untested theory is that this match checks whether `this == Chain.Empty`, which causes an infinite loop due to the `toList` call in `equals`.

Forgive me, but why is it necessarily a problem? Or rather, why does it cause an infinite loop?

Reply:
`this == Chain.Empty` calls `this.toList` which calls `this == Chain.Empty` which calls...

Reply:
Amazing.

Reply:
Wow...

Reply:
That also explains why the OO implementation works.

Reply:
I think we owe Matt a beer.

Reply:
> Can't we sidestep all of this by inheriting `Iterable` and then just doing `Iterable.iterate.toList`?

Hmm... I tried experimenting with this but it doesn't seem like we can do that without using `equals` implicitly somewhere.

Reply:
> > Can't we sidestep all of this by inheriting `Iterable` and then just doing `Iterable.iterate.toList`?
> 
> Hmm... I tried experimenting with this but it doesn't seem like we can do that without using `equals` implicitly somewhere.

Can you hack the `unapply` function to do `instanceof` instead?

Reply:
I'll look into it

Reply:
At the risk of sounding like old man yells at cloud: I dont want to extend Iterable and I don't want to have OO-style.

Reply:
![CleanShot 2024-01-08 at 18 58 52@2x](https://github.com/flix/flix/assets/2287953/afefd42e-9e1a-406b-ae6a-31f8f0bb4481)

Reply:
Why do we need equals on Chain? Maybe thats what we should drop?

Reply:
If we don't redefine it then equals will be structural equality which may be weird.

I disagree with Iterable being bad because it's a very simple interface but I won't fight anyone on it.

IMO just overriding `unapply` will fix everything.

Reply:
We could also drop case classes?